### ISSUE TITLE
Import big CSV files give out of memory error

### ISSUE THREAD

**Issue by qbadx** [memgraph/memgraph#684]
**Memgraph version**
 2.4.2

**Environment**
docker 20.10.12
memgraph official docker image
Ubuntu 22.04.1 LTS
64 GB RAM

**Describe the bug**

Following the procedure to migrate data from neo4j to memgraph (https://memgraph.com/blog/how-to-migrate-from-neo4j-to-memgraph)

I have a 4.2 GB csv file with 36494229 lines and  56 columns out of neo4j but when I try to import it using the LOAD command as specified in the procedure, the used RAM sky rockets and exceed the 64 GB of available RAM.

I tried to split the file in smaller files and somehow managed to make it work with files < 500 MB but it is not scalable

**To Reproduce**

Load a big csv file with the LOAD command

LOAD CSV FROM "/tmp/test.csv" WITH HEADER as row WITH row WHERE ....

**Expected behavior**

Query failed: Memory limit exceeded! Atempting to allocate a chunk of 128B which would put the current use to 65.47GiB, while the maximum allowed size for allocation is set to 62.42GiB.

**Logs**
In memgraph UI
Query failed: Memory limit exceeded! Atempting to allocate a chunk of 128B which would put the current use to 65.47GiB, while the maximum allowed size for allocation is set to 62.42GiB.

In memgraph CLI
mg_raw_transport_recv: connection closed by server
Client received exception: failed to receive chunk size
Trying to reconnect
Connection failure: couldn't connect to host: Connection refused
Connection failure: couldn't connect to host: Connection refused
Connection failure: couldn't connect to host: Connection refused
Couldn't connect to: 'memgraph://127.0.0.1:7687'

**Additional context**

**Comment by qbadx** [comment]
Hello @gitbuda and thx for you reply, 

I can't share the data unfortunately :/

Here are the stats : 

13 node labels and 3.367.565 node IDs
56 property keys and 39.412.325 property IDs
18 relationship types and 33.148.468 relationship IDs

The csv has  3.494.229 lines and 56 columns

Do you need more information ?

**Comment by qbadx** [comment]
update:

I split the big CSV file into several smaller ones based on the label of the nodes and I end up with the biggest file being 500 MB.

This file in particular makes the memory goes very very high (almost 50/64 GB!)

As I said I can't share the data but it contains 2560553 lines with 3 fields : 

- dates YYYY-MM-DD
- strings of 33 alphanumeric characters  ex. 90v728b966f7892c9db0ec8976367176
- URL's 

Hope it helps

**Comment by gitbuda** [comment]
Thanks for the details! I think this is enough + will try to find somebody who can deep dive here 😄 Will keep you posted!

In the meantime, is it an option for you to write a script to transform CSV into Cypher queries and execute them directly against Memgraph? A few notes about that:
* don't forget to create indexes
* node CSV line would translate into `CREATE (n:Label {props});`
* edge CSV line would translate into `MATCH (n1:Label {id_prop: X}) MATCH (n2:Label {id_prop: Y}) CREATE (n1)-[r:TYPE]->(n2);`)
* you can execute directly from e.g. Python or Go, or you can dump all the queries to a file and then use [mgconsole](https://github.com/memgraph/mgconsole)

I think that should work, let me know if you try or if you encounter any issues with that 👀

**Comment by qbadx** [comment]
Hello @gitbuda, 

I have been able to create all the nodes using the technique I mentioned earlier.

I have been using the same technique (LOAD CSV) for the relationships but it is a bit more complicated: 

(Using the UI) Sometimes it works just fine and the relationships are created very quickly and I can visualize them but most of the time I receive an error "Query failed [object ProgressEvent]" and strangely the relationships counter keeps increasing but very very slowly (like 1 relation created every 3 seconds) but it seems that the relationships are actually not created.


Next step is to try what you to create all the relationships using cypher queries...

I keep you posted.

**Comment by antoniofilipovic** [comment]
Hi, @qbadx 

I tried testing your use case now, and there are a few findings I would like to share with you.

Depending on the dataset size, memory can increase. For example, expected memory usage for reading a CSV file with 3.494.229 lines and 56 columns, memory can jump up to: 3.494.229*(56B+56*80B+56*40B) = 23.7GB, depending on how you structure your query. 

Considering your question about importing from a cypherl file, did you follow this structure:


```
CREATE (:`Team` {id: 1, `name`: "A"});
"... create other vertices"
...
CREATE INDEX ON :Team(id);
"create the index by which you will search nodes when creating relationships"
...
MATCH (u:Team), (v:Team) WHERE u.id = 0 AND v.id = 1 CREATE (u)-[:`IS_PART_OF`]->(v);
```
The following structure should drastically speed up the process. Sorry for taking longer than expected to answer the question, but we wanted to test the LOAD CSV command. It seems like there is some memory issue, but we still haven't managed to find the issue. 

Cypherl command should help when importing data.


If you still haven't joined our discord server, [here is the link](https://discord.gg/memgraph) and it would be much easier for us to iterate over the issue you encountered! My nickname there is `afilipovic#8994`

**Comment by antoniofilipovic** [comment]
Hi, just to let you know, we will soon merge PR which should resolve the issue you encountered. If you want, you can check it out [here](https://github.com/memgraph/memgraph/pull/712). The fix should be part of the new Memgraph 2.5.1

**Comment by gitbuda** [comment]
@qbadx the @antoniofilipovic's PR has been merged 💪 I'll leave this open for now, the plan is to release in the following weeks, if you really want to try it out as soon as possible, try to follow this https://memgraph.notion.site/Quick-Start-82a99a85e62a4e3d89f6a9fb6d35626d guide, let me/us know if you stuck somewhere, we'll assist 😄

**Comment by alexdashkov** [comment]
Hello @gitbuda I see that there us a release in github with the fix. However, the `memgraph/memgraph-platform` docker image still uses previous version (`2.5.0` instead of `2.5.1`). Could you provide some information when the new release will be integrated in the `memgraph-platform` as well?

**Comment by gitbuda** [comment]
Hi @alexdashkov! Yes, we had a few minor issues with the platform release, but it is all good now -> https://hub.docker.com/r/memgraph/memgraph-platform/tags 😄 Usually, we intend to release all at once (memgraph, mage, platform, cloud, docs). We already have another release in the pipeline, maybe even next week 😄

I hope your issue god resolved. Let us know about any additional feedback. Happy to help!

**Comment by qbadx** [comment]
Hello, I'm using the 2.5.1 and still has the memory issue. Can't wait for the release of the fix :-)

**Comment by qbadx** [comment]
Hello  @gitbuda @antoniofilipovic , 

I have been trying to process the (same) big csv file using the latest docker image but I still get the memory issue.

Splitting the big csv file into smaller ones works, the biggest csv files is 500 MB and it makes the memory sky rocketing. 

Actually it is exactly the same situation than when I first reported the issue in the first place.

Regards

**Comment by antoniofilipovic** [comment]
@qbadx hi, 

I hope you are doing well. Next week we will release a new Memgraph, where we fixed performance issues others reported also with [LOAD CSV](https://github.com/memgraph/memgraph/pull/753). The next plan is to work on the memory issue, and you can track more of what is happening inside this [PR](https://github.com/memgraph/memgraph/pull/772). 

If you had any other issues, please let us know. And let me know if you had time to try Memgraph as binary instead of Docker image.

**Comment by antoniofilipovic** [comment]
@qbadx Hi once again,

wanted to share with you great news. We released Memgraph 2.7 this week and it comes with a great new feature that should help you massively in importing the dataset inside Memgraph. This feature enables you to change in which mode the storage engine inside Memgraph is operating. You can read more about that here on the docs page about [storage modes](https://memgraph.com/docs/memgraph/reference-guide/storage-modes#analytical-storage-mode). By setting Memgraph to IN_MEMORY_ANALYTICAL storage mode before importing the dataset with LOAD CSV, the import will be much much faster and there would be no memory spikes inside Memgraph. This comes at a small cost as there are no ACID guarantees. The good thing is you can switch back and forth between this IN_MEMORY_ANALYTICAL and IN_MEMORY_TRANSACTIONAL (default one). 

I recommend the following steps when it comes to importing your dataset:

1. Switch Memgraph to IN_MEMORY_ANALYTICAL by running the query: STORAGE MODE IN_MEMORY_ANALYTICAL;
2. Import dataset using [LOAD CSV ](https://memgraph.com/docs/memgraph/import-data/load-csv-clause)cypher clause.
3. Switch back to IN_MEMORY_TRANSACTIONAL with the following query: STORAGE MODE IN_MEMORY_TRANSACTIONAL;
4. Create a snapshot of your data and start running other analytical queries, i.e. BFS, DFS, etc.

I recommend creating a snapshot once you switch back to IN_MEMORY_TRANSACTIONAL as there are no write-ahead logs nor periodic snapshots in IN_MEMORY_ANALYTICAL. Afterward, everything will continue to work as expected and there shouldn't be any issues. In IN_MEMORY_TRANSACTIONAL there are write-ahead logs and periodic snapshots. 

From my testing up until now, you could expect to import 10M nodes in less than 45 seconds, and considering memory, for simple 3 properties (id, and 2 strings up to 10 characters) it would take you around 2.2GB of RAM (before it could take more than 10GB of RAM). 

The further great news is that our team is working on faster snapshot recovery times and other improvements which should be available from Memgraph 2.8. So when you work with IN_MEMORY_ANALYTICAL you can be sure that after importing data and creating a snapshot on the restart, Memgraph will load the dataset really fast. 

Let me know what do you think,

Antonio

**Comment by gitbuda** [comment]
Closing this issue primarily because there is a high probability that the team fixed it + because of inactivity in the last few months, we can reopen it or create a new one if needed 😃
